File: 001. Two Sum/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let result = [0,1]
    let hash_map = new Map() // better than object {}
    nums.forEach( (val, idx) => { // value,index
        if (!(hash_map.has(val))) {
        	hash_map.set(target-val, idx)
        }
        else {
            result = [hash_map.get(val), idx]
        }
    } )
    return result
};

// Runtime 69 ms / Memory 45.3 MB
// debug: https://jsfiddle.net/

File: 001. Two Sum/twoSum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> hashMap; //{value:index}
        int n = nums.size();     

        for (int i=0; i<n; i++) {
            int diff, val;
            val = nums[i];
            diff = target - val;
            if (hashMap.find(val)!=hashMap.end()) { //key exist
                return {hashMap[val], i}; //{diff_index,value_index}
            }
            else {
                hashMap[diff] = i; //{diff:val_index}
            }
        }

        return {};
    }
};